src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs

```diff
@@ -1101,7 +1105,7 @@ public static class Sse2
         /// <summary>
         /// __m128d _mm_sqrt_sd (__m128d a)
         /// </summary>
-        public static Vector128<double> SqrtScalar(Vector128<double> value) => SqrtScalar(value);
+        public static Vector128<double> SqrtScalar(Vector128<double> upper, Vector128<double> value) => SqrtScalar(upper, value);
```

Comment:
From the performance perspective, we always want to duplicate the source register for some of these instructions with VEX-encoding. For example, `vsqrtss xmm0, xmm1, xmm1` is better than  `vsqrtss xmm0, xmm0, xmm1` and `vsqrtss xmm0, xmm2, xmm1`. So `SqrtScalar(upper, value)` may make performance issues that is hard to detected.

Form the user experience perspective, `SqrtScalar(value)` usually is more straightforward and useful than `SqrtScalar(upper, value)`.

In summary, I think we can have  `SqrtScalar(value)` and `SqrtScalar(upper, value)` both if we really need the 2-arg version.

Reply:
> we always want to duplicate the source register for some of these instructions with VEX-encoding.

I think this should say: `The recommendation is to duplicate the source register for some of these instructions.`

I think its one of those things where an analyzer comes in handy (just like the user could do `SetAll(0.0f)`, instead of `SetZero()`, even though the latter is more performant).

>  For example, vsqrtss xmm0, xmm1, xmm1 is better than vsqrtss xmm0, xmm0, xmm1 and vsqrtss xmm0, xmm2, xmm1

However, `vsqrtss target, upper, value` (when upperReg != valueReg) will produce smaller codegen then `movss value, upper, value; vsqrtss target, value, value`, which is what users would need to code otherwise (via `value = MoveScalar(upper, value); target = SqrtScalar(value)`). It is also (theoretically) faster since there are fewer instructions filling the pipeline (the equivalent non-vex instructions behave similarly).

> In summary, I think we can have SqrtScalar(value) and SqrtScalar(upper, value) both if we really need the 2-arg version.

I was in agreement that this could/should be left open as an option. It is not possible to do for the binary scalar simd instructions (like `AddScalar`), but does allow easy exposure of the "recommended" use for the unary scalar simd instructions (with the caveats that it "may" reduce consistency/clarity of the instructions behavior).

Reply:
I agree with @fiigii that good solution would be to have to scalar overloads (i) two arg version and (ii) one arg version

Reply:
> It is not possible to do for the binary scalar simd instructions (like AddScalar), 

Absolutely, I meant just for certain 1-arg SIMD scalar intrinsic, such as `SqrtScalar`, `RoundCurrentDirectionScalar`, `ReciprocalScalar`, etc.

Reply:
I'm fine with the two overloads, though I begin to be a bit concerned about even further exploding the API surface.

Reply:
>  though I begin to be a bit concerned about even further exploding the API surface.

That is one of my concerns as well.

These could possibly be implemented in terms of `SqrtScalar(value) => SqrtScalar(value, value)`, with an aggressive inlining attribute or as a general helper library (CoreFXExtensions) for commonly used algorithms/helpers...

Reply:
> These could possibly be implemented in terms of SqrtScalar(value) => SqrtScalar(value, value), with an aggressive inlining attribute or as a general helper library (CoreFXExtensions) for commonly used algorithms/helpers...

That will make JIT/VM more complex because all the methods in this namespace are automatically labeled with [Intrinsic]. If we want the two versions both, implementing them in JIT via the table-driven framework is really easy.

Reply:
> a bit concerned about even further exploding the API surface

We are much more concise than C++ implementation. The overloads which are proposed are the closest match to original C++ intrinsics and there will be only few of them. This should not have any significant impact on number of methods which would be around 1 300 for current accepted Intel ISA. I have not counted Arm but it will be significantly more concise as there are more methods which can be expressed as generics according to our API design rules.

Reply:
AH, I forgot about the implicit `[Intrinsic]` handling for all methods here